sdk/src/entrypoint/mod.rs

```diff
@@ -595,62 +606,137 @@ unsafe impl GlobalAlloc for NoAllocator {
+        /// Creates the allocator tied to specific range of addresses.
+        ///
+        /// The start address must be aligned to `usize` and the length must be
+        /// at least `size_of::<usize>()` bytes.
+        ///
+        /// # Safety
+        ///
+        /// This is unsafe in most situations, unless you are totally sure that
+        /// the provided start address and length can be written to by the allocator,
+        /// and that the memory will be usable for the lifespan of the allocator.
+        /// The reserved memory region must be at least `size_of::<usize>()` bytes
+        /// long.
+        ///
+        /// For Solana on-chain programs, a certain address range is reserved, so
+        /// the allocator can be given those addresses. In general, the `len` is
+        /// set to the maximum heap length allowed by the runtime. The runtime
+        /// will enforce the actual heap size requested by the program.
+        pub const unsafe fn new_unchecked(start: usize, len: usize) -> Self {
+            Self {
+                start,
+                end: start + len,
+            }
+        }
     }
 
-    /// Integer arithmetic in this global allocator implementation is safe when operating on the
-    /// prescribed [`HEAP_START_ADDRESS`] and [`HEAP_LENGTH`]. Any other use may overflow and is
-    /// thus unsupported and at one's own risk.
+    // Integer arithmetic in this global allocator implementation is safe when operating on the
+    // prescribed `GlobalAlloc::start` and `GlobalAlloc::end`. Any other use may overflow and
+    // is thus unsupported and at one's own risk.
     #[allow(clippy::arithmetic_side_effects)]
     unsafe impl GlobalAlloc for BumpAllocator {
-        /// Allocates memory as a bump allocator.
+        /// Allocates memory as described by the given `layout` using a forward bump allocator.
+        ///
+        /// Returns a pointer to newly-allocated memory, or `null` to indicate allocation failure.
+        ///
+        /// # Safety
+        ///
+        /// `layout` must have non-zero size. Attempting to allocate for a zero-sized layout will
+        /// result in undefined behavior.
         #[inline]
         unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+            // Reads the current position of the heap pointer.
+            //
+            // Integer-to-pointer cast: the caller guarantees that `self.start` is a valid
+            // address for the lifetime of the allocator and aligned to `usize`.
             let pos_ptr = self.start as *mut usize;
-
             let mut pos = *pos_ptr;
-            if pos == 0 {
+
+            if unlikely(pos == 0) {
                 // First time, set starting position.
-                pos = self.start + self.len;
+                pos = self.start + size_of::<usize>();
+            }
+
+            let padding = layout.align() - 1;
+
+            if unlikely(self.end - pos < padding) {
+                return null_mut();
```

Comment:
Maybe I'm being dense, but is this check necessary? `allocation` finds the start address, so we should just need to later check that `self.end > allocation + layout.size()`

Reply:
I was avoiding doing the `allocation + layout.size()` operation without checks since it can overflow if you try to allocate `u64::MAX` for example. And `allocation` might be bigger than `self.end` if there is no "space" left for the alignment padding.

We could probably save `1` CU by testing that `layout.size()` is not greater than `MAX_HEAP_SIZE` and then do the check `self.end > allocation + layout.size()` instead. I will try that.

Reply:
It does save `1` extra CU. 😊